### PR TITLE
ec-586 - optimized cache

### PR DESCRIPTION
**Pull request by GusevTimofey**

### CODE REVIEW

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 1]
```diff
@@ -29,29 +30,35 @@ object ModifiersCache extends StrictLogging {
...
-    }.map(_._1).map(remove)
+    }.map(mod => remove(mod._1))
   }
 
   def remove(key: Key): Option[EncryPersistentModifier] = {
     logger.info(s"Going to delete ${Algos.encode(key.toArray)}. Cache contains: ${cache.get(key).isDefined}")
-    cache.remove(key).map { removed =>
-      removed
-    }
+    cache.remove(key).map(removed => removed)
```
Comment (Bromel777) [inline src/main/scala/encry/view/ModifiersCache.scala]: cache.remove(key)?

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 2]
```diff
@@ -60,17 +67,62 @@ object ModifiersCache extends StrictLogging {
...
+        List[Key]()
+    }
+    if (bestHeadersIds.nonEmpty) bestHeadersIds
+    else if (history.bestHeaderHeight - history.bestBlockHeight > 0) {
+      val payloadId: Option[Key] = history.headerIdsAtHeight(history.bestBlockHeight + 1).headOption match {
+        case Some(id) =>
+          val header: Option[Header] = history.modifierById(id).collect { case header: Header => header }
+          val modFromCache: Option[EncryPersistentModifier] = cache.get {
+            val payloadId: Array[Byte] = header.map(_.payloadId).getOrElse(Array.emptyByteArray)
+            new mutable.WrappedArray.ofByte(payloadId).asInstanceOf[Key]
```
Comment (Bromel777) [inline src/main/scala/encry/view/ModifiersCache.scala]: remove .asInstanceOf[Key]

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 3]
```diff
@@ -60,17 +67,62 @@ object ModifiersCache extends StrictLogging {
...
+    val bestHeadersIds: List[Key] = headerIds match {
+      case Some(value) =>
+        headersCollection = headersCollection - (history.bestHeaderHeight + 1)
+        logger.info(s"HeadersCollection size is: ${headersCollection.size}")
+        value.map(m => cache.get(m)).collect {
+          case Some(v: Header)
+            if (v.parentId sameElements history.bestHeaderOpt.map(_.id).getOrElse(Array.emptyByteArray))
+              && isApplicable(v.id) =>
+            logger.info(s"Find new bestHeader in cache: ${Algos.encode(v.id)}")
+            new mutable.WrappedArray.ofByte(v.id).asInstanceOf[Key]
```
Comment (Bromel777) [inline src/main/scala/encry/view/ModifiersCache.scala]: remove .asInstanceOf[Key]

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 4]
```diff
@@ -29,29 +30,35 @@ object ModifiersCache extends StrictLogging {
 
   def put(key: Key, value: EncryPersistentModifier, history: EncryHistory): Unit = if (!contains(key)) {
     cache.put(key, value)
+    value match {
+      case header: Header =>
+        val currentHeight: Block.Height = header.height
+        val possibleHeadersAtCurrentHeight: List[ModifierId] = headersCollection.getOrElse(currentHeight, List())
```
Comment (Bromel777) [inline src/main/scala/encry/view/ModifiersCache.scala]: replace currentHeight to header.height

File: src/main/scala/encry/view/ModifiersCache.scala [hunk 5]
```diff
@@ -29,29 +29,34 @@ object ModifiersCache extends StrictLogging {
...
         case Success(_) => false
         case Failure(_: RecoverableModifierError) => false
         case _ => true
       }
-    }.map(_._1).map(remove)
+    }.map(mod => remove(mod._1))
   }
 
   def remove(key: Key): Option[EncryPersistentModifier] = {
     logger.info(s"Going to delete ${Algos.encode(key.toArray)}. Cache contains: ${cache.get(key).isDefined}")
```
Comment (GRYE) [inline src/main/scala/encry/view/ModifiersCache.scala]: Do we need this log message?

**Pull request merged**